### DISCUSSION TITLE
Using a tRPC server from other langauges/applications.

### DISCUSSION THREAD

**Discussion by bbuck** [trpc/trpc#4697] [General]
So I saw in the documentation (specifically this page https://trpc.io/docs/further-reading) this comment:

> If you already work in a team where languages are mixed or have third-party consumers over whom you have no control, you should create a language-agnostic [GraphQL](https://graphql.org/)-API.

Specifically the codebase we're working with has a web platform in JS (right now) using standard Express with a React front-end. And we have a Unity-built Native/Desktop Application too (so written in C#). The existing API is all over the place and severely bloated. As part of an advancement in building out new features we've decided we're going to reset and build a v2 API alongside the existing on but we're not sure on what structure that API will have.

Having previously used RPC through protobufs and gRPC in the past, tooling outside of certain corps is less than ideal which makes working with those technologies a bit more painful so I was drawn to tRPC which seems to provide a really simplistic approach that's easy to work with and also seems incredibly dev friendly (thinking about onboarding new devs on the project). But again, we have to support a C# SDK that calls the platform too -- but I don't want to go through the pain points of getting GraphQL implemented just to do something like that.

Does the tRPC team/community recommend against using tRPC with any clients other than a tightly coupled TypeScript client? I see that the call structure is JSON-RPC (ish, at least, not familiar with JSON-RPC honestly) and so it seems entirely reasonable we could build our own C# SDK and call the API just fine but I'm hesitant because that line in the docs that says, "go with something else" (paraphrased) if you've got multiple languages.

Just trying to see what other folks have done, think, etc... if you have a working multi-language project using tRPC in prod I'd love to hear those thoughts and opinions as well!

**Comment by ryneex** [comment]
@bbuck  Can't believe it got no reply since I had the same question in mind, so what did you guys end up doing?

**Reply by bbuck** [reply]
Nothing special, honestly. Just straight up HTTP calls. It can be a bit awkward/painful but it's not that bad actually.

Query procedures map to `GET <endpoint>?input={}` where the query parameter input is the JSON string (no wrapping quotes).

Mutation procedures map to `POST <endpoint>` and the body of the request should be the JSON input object.

Unless you're using batch queries, then it gets a bit more complicated -- you can always inspect your network tab and just grab the queries there.

Is this the best solution? Absolutely not -- but it worked for us.

**Comment by devidw** [comment]
One could use https://github.com/jlalmes/trpc-openapi and generate a c# rest api client based on the openapi doc from the trpc service, tho you would be doing rest/openapi then and not really trpc anymore

Note: you have to explicitly write input and output validators for all procedures in your trpc codebase to be able to generate the openapi doc

**Reply by bbuck** [reply]
Yea, I just recently saw that. I can kind of understand why it exists but it absolutely doesn't suit our needs at all. We have multiple tRPC services, so configuring them all to be consolidated into a single rest server that we then end up having to host in addition to all the tRPC sources is just a huge pain. I'm sure this use case works for others, I know in the past when I used gRPC we also had an external REST<->gRPC mapping but, again, that doesn't suit our needs where I'm at now.

We just went with making the raw requests. It's a bit awkward but it works fine.

**Reply by devidw** [reply]
oh u don't need to put them behind a single rest server

u could continue to have u sep trpc services, and each just exposes a rest server alongside the trpc server

the only reason that would make sense imo is that u can auto-gen a client, thus get type-safety with your client

tho at the price of having to maintain input & output validators in trpc

which is especially annoying when u return deeply nested ts types

**Reply by whkelvin** [reply]
Not sure if this will fit your needs, but I am developing a new trpc-like and language-agnostic protocol based on http POST requests. It's still super early stage, but you can get an idea of what I am trying to build here: https://yizy.rootxsnowstudio.com/demo
